File: tree/0687_longest_univalue_path/0687_longest_univalue_path.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def longestUnivaluePath(self, root):
        result = [0]
        def dfs(node):
            if not node:
                return 0
            left, right = dfs(node.left), dfs(node.right)
            if node.left and node.left.val == node.val:
                left = left + 1
            else:
                left = 0
            if node.right and node.right.val == node.val:
                right = right + 1
            else:
                right = 0
            result[0] = max(result[0], left + right)
            return max(left, right)
        
        dfs(root)
        return result[0]

root = TreeNode(1)
root.left = TreeNode(1)
root.right = TreeNode(1)
res = Solution().longestUnivaluePath(root)
print(res)

File: tree/0687_longest_univalue_path/0687_longest_univalue_path.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func longestUnivaluePath(root *TreeNode) int {
	maxLen := 0
	helper(root, &maxLen)
	return maxLen
}

func helper(n *TreeNode, maxLen *int) int {
	if n == nil {
		return 0
	}
	
	l := helper(n.Left, maxLen)
	r := helper(n.Right, maxLen)
	res := 0

	if n.Left != nil && n.Val == n.Left.Val {
		*maxLen = max(*maxLen, l + 1)
		res = max(res, l+1)
	}

	if n.Right != nil && n.Val == n.Right.Val {
		*maxLen = max(*maxLen, r + 1)
		res = max(res, r + 1)
	}

	if n.Left != nil && n.Val == n.Left.Val && n.Right != nil && n.Val == n.Right.Val {
		*maxLen = max(*maxLen, l + r + 2)
	}

	return res
}

func max(a,b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	root := &TreeNode{1, nil,nil}
	root.Left = &TreeNode{1, nil,nil}
	root.Right = &TreeNode{1, nil,nil}
	res := longestUnivaluePath(root)
	fmt.Println(res)
}